services/src/main/java/org/keycloak/broker/saml/mappers/XPathAttributeMapper.java

```diff
@@ -0,0 +1,316 @@
+                return "";
+            } catch (Exception e) {
+                throw new RuntimeException("Could not parse xml element", e);
+            }
+        };
+    }
+
+    private List<String> findAttributeValuesInContext(String attributeName, String attributeXPath, BrokeredIdentityContext context) {
+        AssertionType assertion = (AssertionType) context.getContextData().get(SAMLEndpoint.SAML_ASSERTION);
+
+        return assertion.getAttributeStatements().stream()
+                .flatMap(statement -> statement.getAttributes().stream())
+                .filter(elementWith(attributeName))
+                .flatMap(attributeType -> attributeType.getAttribute().getAttributeValue().stream())
+                .filter(String.class::isInstance)
+                .map(String.class::cast)
+                .map(s -> "<root>" + s + "</root>")
+                .map(applyXPath(attributeXPath))
+                .filter(Objects::nonNull)
+                .map(Object::toString)
+                .filter(x -> !x.isEmpty())
+                .collect(Collectors.toList());
+    }
+
+    @Override
+    public void updateBrokeredUser(KeycloakSession session, RealmModel realm, UserModel user, IdentityProviderMapperModel mapperModel, BrokeredIdentityContext context) {
+        String attribute = mapperModel.getConfig().get(USER_ATTRIBUTE);
+        if (StringUtil.isNullOrEmpty(attribute)) {
+            return;
+        }
+        String attributeName = getAttributeNameFromMapperModel(mapperModel);
+        String attributeXPath = mapperModel.getConfig().get(ATTRIBUTE_XPATH);
+        List<String> attributeValuesInContext = findAttributeValuesInContext(attributeName, attributeXPath, context);
+        if (attribute.equalsIgnoreCase(EMAIL)) {
+            setIfNotEmptyAndDifferent(user::setEmail, user::getEmail, attributeValuesInContext);
+        } else if (attribute.equalsIgnoreCase(FIRST_NAME)) {
+            setIfNotEmptyAndDifferent(user::setFirstName, user::getFirstName, attributeValuesInContext);
+        } else if (attribute.equalsIgnoreCase(LAST_NAME)) {
+            setIfNotEmptyAndDifferent(user::setLastName, user::getLastName, attributeValuesInContext);
+        } else {
+            List<String> currentAttributeValues = user.getAttributes().get(attribute);
+            if (attributeValuesInContext == null) {
+                // attribute no longer sent by brokered idp, remove it
+                user.removeAttribute(attribute);
+            } else if (currentAttributeValues == null) {
+                // new attribute sent by brokered idp, add it
+                user.setAttribute(attribute, attributeValuesInContext);
+            } else if (!CollectionUtil.collectionEquals(attributeValuesInContext, currentAttributeValues)) {
+                // attribute sent by brokered idp has different values as before, update it
+                user.setAttribute(attribute, attributeValuesInContext);
+            }
+            // attribute already set
```

Comment:
Please replace with solely attribute manipulation

Reply:
@hmlnarik - I assume that you're referring to the following logic in the Keycloak model to map attributes to fields, like for example here: https://github.com/keycloak/keycloak/blob/main/model/jpa/src/main/java/org/keycloak/models/jpa/UserAdapter.java#L161-L183

With this approach, the cases for EMAIL, FIRST_NAME, and LAST_NAME can be removed, and only the else-branch is necessary. I gave it a try with the new SAML test cases, and it works as expected. 

At the same time [BrokeredIdentityContext](https://github.com/keycloak/keycloak/blob/main/server-spi-private/src/main/java/org/keycloak/broker/provider/BrokeredIdentityContext.java) doesn't implement this logic, and for example the email stored in an attribute might be hidden from those calling getEmail().

WDYT?

Reply:
> I assume that you're referring to the following logic in the Keycloak model to map attributes to fields, like for example here: https://github.com/keycloak/keycloak/blob/main/model/jpa/src/main/java/org/keycloak/models/jpa/UserAdapter.java#L161-L183
>
> At the same time [BrokeredIdentityContext](https://github.com/keycloak/keycloak/blob/main/server-spi-private/src/main/java/org/keycloak/broker/provider/BrokeredIdentityContext.java) doesn't implement this logic, and for example the email stored in an attribute might be hidden from those calling getEmail().

You are correct, `BrokeredIdentityContext` has been forgotten in the #7080 PR. Thank you for checking this. Let's keep it as is for now, and address it in the newly introduced #12307.

Reply:
If I understood you correctly I should nevertheless change it already in my code? I replaced it including the else path (@ahus1) with a call to `setUserAttribute` which should be working as expected as far as I understand the code. Let me know if that was wrong / is not the desired change. I also had to change the unittest ... so this changed some behaviour.

Reply:
> If I understood you correctly I should nevertheless change it already in my code?

Yes, that's correct, thank you.